File: dotnet/rpc/xdr.cs

using System;

namespace fju {
  public class XDR {
    public byte[] buf;
    public int count;
    public int offset;

    public XDR(byte[] buf) {
      this.buf = buf;
      this.count = buf.Length;
      this.offset = 0;
    }

    public XDR(int count) {
      this.buf = new byte[count];
      this.count = count;
      this.offset = 0;
    }

    public void encodeUInt32(UInt32 x) {
      buf[offset++] = (byte)((x >> 24) & 0xff);
      buf[offset++] = (byte)((x >> 16) & 0xff);
      buf[offset++] = (byte)((x >> 8) & 0xff);
      buf[offset++] = (byte)((x) & 0xff);
    }

    public UInt32 decodeUInt32() {
       UInt32 x = 0;
       x = buf[offset++];
       x = (x << 8) | buf[offset++];
       x = (x << 8) | buf[offset++];
       x = (x << 8) | buf[offset++];
       return x;
    }

    public void encodeString(String x) {
      byte[] octets = System.Text.Encoding.UTF8.GetBytes(x);
      encodeOpaque(octets);
    }

    public String decodeString() {
      return System.Text.Encoding.UTF8.GetString(decodeOpaque());
    }
 
    public void encodeOpaque(byte[] opaque) {
      encodeUInt32((UInt32)opaque.Length);
      for(int i = 0; i < opaque.Length; i++ ) {
        buf[offset++] = opaque[i];
      }
      if( opaque.Length % 4 != 0 ) {
        offset += 4 - (int)(opaque.Length % 4);
      }
    }

    public byte[] decodeOpaque() {
      UInt32 len = decodeUInt32();
      byte[] octets = new byte[len];
      for( int i = 0; i < len; i++ ) {
        octets[i] = buf[offset++];
      }
      if( len % 4 != 0 ) {
        offset += (4 - (int)(len % 4));
      }
       return octets;
    }

    public void encodeBoolean(bool x) {
      encodeUInt32((UInt32)(x ? 1 : 0));
    }
    public bool decodeBoolean() {
      UInt32 x = decodeUInt32();
      return x == 0 ? false : true;
    }
    
  }
}

File: go/fju/xdr/xdr.go
package xdr

type Xdr struct {
	offset int
	buf []byte
}

func New(buf []byte) *Xdr {
	return &Xdr {
		buf: buf,
	}
}

func (xdr *Xdr) Reset() {
	xdr.offset = 0
}

func (xdr *Xdr) Offset() int {
	return xdr.offset
}

func (xdr *Xdr) Buffer() []byte {
	return xdr.buf[0:xdr.offset]
}

func (xdr *Xdr) EncodeUInt32(x uint32) {
	xdr.buf[xdr.offset] = byte(x >> 24)
	xdr.buf[xdr.offset + 1] = byte(x >> 16)
	xdr.buf[xdr.offset + 2] = byte(x >> 8)
	xdr.buf[xdr.offset + 3] = byte(x)
	xdr.offset += 4
}

func (xdr *Xdr) DecodeUInt32() uint32 {
	var x uint32
	x = 0
	xdr.offset += 4
	return x
}

func (xdr *Xdr) EncodeInt(x int) {
	xdr.EncodeUInt32(uint32(x))
}

func (xdr *Xdr) DecodeInt() int {
	return int(xdr.DecodeUInt32())
}

func (xdr *Xdr) EncodeUInt64(x uint64) {
	xdr.EncodeUInt32(uint32(x >> 32))
	xdr.EncodeUInt32(uint32(x))
}

func (xdr *Xdr) DecodeUInt64() uint64 {
	h := xdr.DecodeUInt32()
	l := xdr.DecodeUInt32()
	return (uint64(h) << 32) | uint64(l)
}

func (xdr *Xdr) EncodeString(x string) {
	len := len(x)
	xdr.EncodeInt(len)
	xdr.EncodeFixed([]byte(x))
}

func (xdr *Xdr) DecodeString() string {
	len := xdr.DecodeInt()
	buf := xdr.buf[xdr.offset:xdr.offset + len]
	xdr.offset += len
	if len % 4 != 0 {
		xdr.offset += 4 - (len % 4)
	}
	return string(buf)
}

func (xdr *Xdr) EncodeOpaque(x []byte) {
	xdr.EncodeInt(len(x))
	copy(xdr.buf[xdr.offset:], x)
	xdr.offset += len(x)
	if len(x) % 4 != 0 {
		xdr.offset += 4 - (len(x))
	}
}

func (xdr *Xdr) DecodeOpaque() []byte {
	len := xdr.DecodeInt()
	buf := make([]byte, len)
	copy(buf, xdr.buf[xdr.offset:xdr.offset + len])
	xdr.offset += len
	if len % 4 != 0 {
		xdr.offset += 4 - (len % 4)
	}
	return buf
}

func (xdr *Xdr) EncodeFixed(x []byte) {
	copy(xdr.buf[xdr.offset:], x)
	xdr.offset += len(x)
	if len(x) % 4 != 0 {
		xdr.offset += 4 - (len(x) % 4)
	}
}

func (xdr *Xdr) DecodeFixed(len int) []byte {
	buf := make([]byte, len)
	copy(buf, xdr.buf[xdr.offset:xdr.offset + len])
	xdr.offset += len
	if len % 4 != 0 {
		xdr.offset += 4 - (len % 4)
	}
	return buf
}



	
	


